From: Michael Biebl <biebl@debian.org>
Date: Thu, 18 Jul 2013 20:11:02 +0200
Subject: timedated: also store timezone configuration in /etc/timezone

Forwarded: not-needed
---
 src/timedate/timedated.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/timedate/timedated.c b/src/timedate/timedated.c
index 204031f..79aa6b9 100644
--- a/src/timedate/timedated.c
+++ b/src/timedate/timedated.c
@@ -136,6 +136,10 @@ static int context_read_data(Context *c) {
                 }
         }
 
+        r = read_one_line_file("/etc/timezone", &c->zone);
+        if (r < 0 && r != -ENOENT)
+                log_warning("Failed to read /etc/timezone: %s", strerror(-r));
+
 have_timezone:
         if (isempty(c->zone)) {
                 free(c->zone);
@@ -150,6 +154,7 @@ have_timezone:
 static int context_write_data_timezone(Context *c) {
         _cleanup_free_ char *p = NULL;
         int r = 0;
+        struct stat st;
 
         assert(c);
 
@@ -157,6 +162,9 @@ static int context_write_data_timezone(Context *c) {
                 if (unlink("/etc/localtime") < 0 && errno != ENOENT)
                         r = -errno;
 
+                if (unlink("/etc/timezone") < 0 && errno != ENOENT)
+                        r = -errno;
+
                 return r;
         }
 
@@ -168,6 +176,10 @@ static int context_write_data_timezone(Context *c) {
         if (r < 0)
                 return r;
 
+        r = write_string_file_atomic("/etc/timezone", c->zone);
+        if (r < 0)
+                return r;
+
         return 0;
 }
 
